senza/respawn.py

```diff
@@ -0,0 +1,125 @@
+
+
+def get_instances_in_service(group, region: str):
+    '''Get set of instance IDs with ELB "InService" state'''
+    instances_in_service = set()
+    # TODO: handle auto scaling groups without any ELB
+    lb_names = group['LoadBalancerNames']
+    if lb_names:
+        # check ELB status
+        elb = boto3.client('elb')
+        for lb_name in lb_names:
+            result = elb.describe_instance_health(LoadBalancerName=lb_name)
+            for instance in result['InstanceStates']:
+                if instance['State'] == 'InService':
+                    instances_in_service.add(instance['InstanceId'])
+    else:
+        # just use ASG LifecycleState
+        group = get_auto_scaling_group(boto3.client('autoscaling'), group['AutoScalingGroupName'])
+        for instance in group['Instances']:
+            if instance['LifecycleState'] == 'InService':
+                instances_in_service.add(instance['InstanceId'])
+    return instances_in_service
+
+
+def scale_out(asg, asg_name, region: str, new_min_size, new_max_size, new_desired_capacity: int):
+    '''Scale out the given Auto Scaling Group to the given capacity
+    and wait for all instances to become "InService" in ELB'''
+    with Action('Scaling to {} instances..'.format(new_desired_capacity)) as act:
+        asg.update_auto_scaling_group(AutoScalingGroupName=asg_name,
+                                      MinSize=new_min_size,
+                                      MaxSize=new_max_size,
+                                      DesiredCapacity=new_desired_capacity)
+        while True:
+            current_group = get_auto_scaling_group(asg, asg_name)
+            instances_in_service = get_instances_in_service(current_group, region)
+            if len(instances_in_service) >= new_desired_capacity:
+                break
+            time.sleep(5)
+            act.progress()
+    return current_group
+
+
+def terminate_instance(asg, region: str, group, instance: str):
+    '''Terminate a single EC2 instance in given Auto Scaling Group and wait for it to become "OutOfService"'''
+
+    with Action('Terminating old instance {}..'.format(instance)) as act:
+        asg.terminate_instance_in_auto_scaling_group(InstanceId=instance, ShouldDecrementDesiredCapacity=False)
+        instances_in_service = get_instances_in_service(group, region)
+        while instance in instances_in_service:
+            time.sleep(2)
+            act.progress()
+            instances_in_service = get_instances_in_service(group, region)
+
+
+def do_respawn_auto_scaling_group(asg_name: str, group: dict, region: str,
+                                  instances_to_terminate: set, instances_ok: set):
+    asg = boto3.client('autoscaling', region)
+    with Action('Suspending scaling processes for {}..'.format(asg_name)):
+        asg.suspend_processes(AutoScalingGroupName=asg_name, ScalingProcesses=SCALING_PROCESSES_TO_SUSPEND)
+    new_min_size = group['MinSize'] + 1
```

Comment:
OK, but what would be the behavior for "--inplace" with only 1 instance? This would mean having a downtime (treating "--inplace" and 1 instance specially would be unexpected/inconsistent).

I can live with having an "--inplace" option, but then it means it's really dangerous for "unaware" users.